### ISSUE TITLE
Carthage with .gitmodule file

### ISSUE THREAD

**Issue by scottphc** [DenHeadless/DTModelStorage#20]
If .gitmodule exist, run `carthage update` will rebuild the whole realm library. It will spend a long time. I try to delete .gitmodule and commit it, and then `carthage update` again. The command will show `*** Downloading realm-cocoa.framework binary at "v2.8.3"`. Very quickly.

So just want to ask why you want to add .gitmodule? Thanks.

**Comment by DenHeadless** [comment]
Hey!

This was a part of Apple's recommendation to always build from source on Swift 2 and Swift 3. I wanted to revisit project setup after ABI-stability was reached, but this never happened =) 

Second reason was to simplify CI, i did not want to call `carthage update` or `carthage bootstrap` on every CI job.

That said, I do want to make use of prebuilt frameworks in the future, and current project setup is definitely temporary. If #19 will be successful, i will probably revisit project setup and migrate to prebuilt frameworks, as they are much simpler than using git submodules.

**Comment by scottphc** [comment]
Just let you know. if use DTCollectionViewManager and DTTableViewManager at the same time, so it will build realm twice.

**Comment by DenHeadless** [comment]
Understood. That is definitely something I want to avoid, so making use of prebuilt frameworks is important.

**Comment by DenHeadless** [comment]
@scottphc It seems I was able to reconfigure repositories to ditch git submodules in favor of carthage setup with the help of travis folder caching and github access tokens.

Can I ask you to try the new setup out?

DTModelStorage, DTCollectionViewManager and DTTableViewManager all have branch `experiment_prebuilt_frameworks`, that you can use in your cartfile to build those frameworks. Currently only `Realm` has a prebuilt binary, but even this should cut your build times a lot. If this setup works on your side, I will merge it to master branches and make new releases.

**Comment by scottphc** [comment]
Hey,

Today I tried DTModelStorage, DTCollectionViewManager, DTTableViewManager with `experiment_prebuilt_frameworks` and got the below error: 

```
CompileSwift normal armv7
    cd /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift -frontend -c /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/SupplementaryMappingTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/SupplementaryEventsTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/NibHeaderFooterView.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/NiblessHeaderFooterView.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/StoryboardCollectionViewCell.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/CollectionViewFactoryTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/StoryboardMappingTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/FixtureCells.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/CreationTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/StoryboardViewController.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/ReactingHeaderFooterView.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Fixtures/XibCollectionViewController.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/DataSourceTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/DelegateForwardingTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Helpers/CollectionViewManager+UnitTests.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/VIewModelMappingCustomizableTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Helpers/DTTestCollectionController.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/StoryboardCollectionReusableView.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/Helpers/NSIndexPath+Convenience.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/ReactingToEventsTestCase.swift /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/MappingTestCase.swift -target armv7-apple-ios8.0 -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS10.3.sdk -I /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Products/Release-iphoneos -F /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Products/Release-iphoneos -F /Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Carthage/Build/iOS -F /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/Library/Frameworks -enable-testing -g -module-cache-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/ModuleCache -serialize-debugging-options -serialize-debugging-options -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/swift-overrides.hmap -Xcc -iquote -Xcc /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Tests-iOS-generated-files.hmap -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Tests-iOS-own-target-headers.hmap -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Tests-iOS-all-non-framework-target-headers.hmap -Xcc -ivfsoverlay -Xcc /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/all-product-headers.yaml -Xcc -iquote -Xcc /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Tests-iOS-project-headers.hmap -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Products/Release-iphoneos/include -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/DerivedSources/armv7 -Xcc -I/Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/DerivedSources -Xcc -working-directory/Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager -emit-module-doc-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/Tests_iOS.swiftdoc -serialize-diagnostics-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/SupplementaryMappingTestCase.dia -O -module-name Tests_iOS -emit-module-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/Tests_iOS.swiftmodule -emit-objc-header-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/Tests_iOS-Swift.h -emit-dependencies-path /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/SupplementaryMappingTestCase.d -num-threads 4 -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/SupplementaryMappingTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/SupplementaryEventsTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/NibHeaderFooterView.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/NiblessHeaderFooterView.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/StoryboardCollectionViewCell.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/CollectionViewFactoryTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/StoryboardMappingTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/FixtureCells.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/CreationTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/StoryboardViewController.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/ReactingHeaderFooterView.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/XibCollectionViewController.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/DataSourceTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/DelegateForwardingTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/CollectionViewManager+UnitTests.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/VIewModelMappingCustomizableTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/DTTestCollectionController.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/StoryboardCollectionReusableView.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/NSIndexPath+Convenience.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/ReactingToEventsTestCase.o -o /Users/scott/Library/Caches/org.carthage.CarthageKit/DerivedData/DTCollectionViewManager/7f728fd0b075481bae9fe5650baa9caeef6d716b/Build/Intermediates/DTCollectionViewManager.build/Release-iphoneos/Tests-iOS.build/Objects-normal/armv7/MappingTestCase.o
/Users/scott/Developer/Test/carthage_test/Carthage/Checkouts/DTCollectionViewManager/Tests/SupplementaryMappingTestCase.swift:10:8: error: no such module 'Nimble'
import Nimble
       ^

** BUILD FAILED **


The following build commands failed:
	CompileSwift normal arm64
	CompileSwiftSources normal arm64 com.apple.xcode.tools.swift.compiler
	CompileSwiftSources normal armv7 com.apple.xcode.tools.swift.compiler
	CompileSwift normal armv7
(4 failures)
```

**Comment by DenHeadless** [comment]
Hmmmm. I don't understand, those are test files from DTCollectionViewManager, that are not included in main targets. And Nimble framework, which is used in unit tests, is listed in Cartfile.private, according to Carthage recommendations.

I dont know what is going on here, sorry. Unfortunately i don't have time to investigate this more at the moment... I would really appreciate your help if you have time to research why it behaves like so.

**Comment by DenHeadless** [comment]
Hey, @scottphc!

I released 6.0.0-beta.1 release, which is now providing prebuilt frameworks for `DTModelStorage`, `DTTableViewManager` and `DTCollectionViewManager`, also dropping `Realm` dependency, which should cut your install times a lot. This release also fixes compilation issues mentioned here. Keep in mind however, that prebuilt binaries will be built only using latest compiler(which is Xcode 9 beta 6 as of this moment). 

Closing.
